app-text/vale/vale-3.5.0.ebuild

```
@@ -0,0 +1,45 @@
+# Copyright 2021-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit go-module optfeature
+
+DESCRIPTION="A syntax-aware linter for prose built with speed and extensibility in mind"
+HOMEPAGE="https://vale.sh/docs/"
+
+SRC_URI="
+	https://github.com/errata-ai/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz
```

```suggestion
	https://github.com/errata-ai/vale/archive/v${PV}.tar.gz -> ${P}.tar.gz
```
Minor nitpick: https://devmanual.gentoo.org/ebuild-writing/variables/index.html#use-of-constant-value-variables-in-ebuilds

---

_Hello @antecrescent,_

Thank you for the review!

I've checked the devmanual and it only suggest to avoid using `${PN}` in the **domain** of `SRC_URI`.
> The examples of bad constant-value variable references are:
>   * using `${PN}` in `HOMEPAGE`, `EGIT_REPO_URI` or the domain of `SRC_URI` — it breaks URL parsing in editors and terminals, and makes it hard to copy the link for external use (e.g. while reviewing via gitweb), 

Meanwhile, using `${PN}` to render the rest of `SRC_URI` is the common practice, so I believe we can keep it as is.
```shell
gentoo (master|✔) $ ag 'SRC_URI=.*/\${PN}/' | wc -l
5677
```

---

Yes, you are correct. I personally prefer not using `${PN}` in the repository part of the URL for the exact reason outlined above:
> it breaks URL parsing in editors and terminals, and makes it hard to copy the [repository] link for external use

I don't see how `SRC_URI="https://github.com/errata-ai/${PN}/...` would be acceptable but `EGIT_REPO_URI="https://github.com/errata-ai/${PN}.git` would not be, when the reason given is that it breaks URL (!) parsing in both cases.

---

That grep doesn't tell the real story. Check instead to see how many packages do this:

```
https://github.com/errata-ai/${PN}/releases/download/v${PV}/${PN}-v${PV}.tar.gz
```

Rather than doing
```
https://github.com/errata-ai/vale/releases/download/v${PV}/${PN}-v${PV}.tar.gz
```

In the latter case, PN is still being used but only for the tarball name and is explicitly avoided for the domain (which by the spirit of the law isn't just the thing with an IP address, but the entire portion of the url that is dedicated to the project -- in the case of a GitHub url, including the organization + repository.

---

_Hello @eli-schwartz,_

Sorry, I'm not sure if I get the point, but thank you for the input!
Do you say it's worth replacing `${PN}` with `vale` here?

> That grep doesn't tell the real story.

That regular expression only matches lines where `${PN}` is surrounded with slashes, so I believe it corresponds to your first example.
<details><summary>gentoo (master|✔) $ ag 'SRC_URI=.*/\${PN}/' | head</summary>
<p>

```text
dev-go/get-ego-vendor/get-ego-vendor-0.2.0.ebuild:9:SRC_URI="https://github.com/williamh/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
x11-wm/amiwm/amiwm-0.22.01_p20230916.ebuild:11:SRC_URI="https://github.com/chewi/${PN}/archive/${COMMIT}.tar.gz -> ${P}.tar.gz"
x11-wm/ratpoison/ratpoison-1.4.9-r3.ebuild:10:SRC_URI="https://download.savannah.gnu.org/releases/${PN}/${P}.tar.xz"
x11-wm/dwm/dwm-6.5.ebuild:15:	SRC_URI="https://dl.suckless.org/${PN}/${P}.tar.gz"
x11-wm/dwm/dwm-9999.ebuild:15:	SRC_URI="https://dl.suckless.org/${PN}/${P}.tar.gz"
x11-wm/matwm2/matwm2-0.1.2_pre3-r2.ebuild:12:SRC_URI="https://github.com/segin/${PN}/archive/${MY_PV}.tar.gz -> ${P}.tar.gz"
x11-wm/jwm/jwm-2.4.2.ebuild:10:SRC_URI="https://github.com/joewing/${PN}/releases/download/v${PV}/${P}.tar.xz"
x11-wm/enlightenment/enlightenment-0.26.0.ebuild:10:SRC_URI="https://download.enlightenment.org/rel/apps/${PN}/${P}.tar.xz"
x11-wm/echinus/echinus-0.4.9.ebuild:10:SRC_URI="https://plhk.ru/static/${PN}/${P}.tar.gz"
x11-wm/notion/notion-4.0.2_p20220613-r1.ebuild:19:	SRC_URI="https://github.com/raboof/${PN}/archive/${COMMIT}.tar.gz -> ${P}.tar.gz"
```

</p>
</details>

A bit more stats for [Gentoo ebuild repository](https://github.com/gentoo/gentoo):
```shell
gentoo (master|✔) $ ag 'SRC_URI=.*/\${PN}/' | wc -l  # only lines like https://github.com/errata-ai/${PN}/archive/v${PV}.tar.gz
5677
gentoo (master|✔) $ ag 'SRC_URI=.*\${PN}' | wc -l  # total
6340
```

As you can see, 89.54% of all cases (5677 out of 6340) for straight-line `SRC_URI`s that use `PN` are similar to what we have in the proposed ebuild.

A bit more findings here:

```shell
gentoo (master|✔) $ ag 'SRC_URI=.*/\${PN}/archive/' | wc -l
2666
```

<details><summary>gentoo (master|✔) $ ag 'SRC_URI=.*/\${PN}/archive/' | head</summary>
<p>

```text
dev-go/get-ego-vendor/get-ego-vendor-0.2.0.ebuild:9:SRC_URI="https://github.com/williamh/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
x11-wm/amiwm/amiwm-0.22.01_p20230916.ebuild:11:SRC_URI="https://github.com/chewi/${PN}/archive/${COMMIT}.tar.gz -> ${P}.tar.gz"
x11-wm/matwm2/matwm2-0.1.2_pre3-r3.ebuild:12:SRC_URI="https://github.com/segin/${PN}/archive/${MY_PV}.tar.gz -> ${P}.tar.gz"
x11-wm/matwm2/matwm2-0.1.2_pre3-r2.ebuild:12:SRC_URI="https://github.com/segin/${PN}/archive/${MY_PV}.tar.gz -> ${P}.tar.gz"
x11-wm/notion/notion-4.0.2_p20220613-r1.ebuild:19:	SRC_URI="https://github.com/raboof/${PN}/archive/${COMMIT}.tar.gz -> ${P}.tar.gz"
x11-wm/notion/notion-4.0.2_p20220613.ebuild:19:	SRC_URI="https://github.com/raboof/${PN}/archive/${COMMIT}.tar.gz -> ${P}.tar.gz"
x11-wm/notion/notion-9999.ebuild:17:	SRC_URI="https://github.com/raboof/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz"
x11-wm/lumina/lumina-1.6.2-r1.ebuild:11:SRC_URI="https://github.com/trueos/${PN}/archive/v${PV/_/-}.tar.gz -> ${P}.tar.gz"
x11-wm/lumina/lumina-1.6.2.ebuild:11:SRC_URI="https://github.com/trueos/${PN}/archive/v${PV/_/-}.tar.gz -> ${P}.tar.gz"
sys-apps/lm-sensors/lm-sensors-3.6.0-r1.ebuild:13:	SRC_URI="https://github.com/lm-sensors/${PN}/archive/${COMMIT}.tar.gz -> ${P}.tar.gz"
```

</p>
</details>

---

@Jamim if we go back to the dev manual,
>

    The examples of bad constant-value variable references are:

        using ${PN} in HOMEPAGE, EGIT_REPO_URI or the domain of SRC_URI — it breaks URL parsing in editors and terminals, and makes it hard to copy the link for external use (e.g. while reviewing via gitweb),

Focus on this part:
>makes it hard to copy the link for external use (e.g. while reviewing via gitweb),

When you have ${PN} in the URI this way, and want to visit the GitHub repo, you'll have to manually change the ${PN} to the package name. Hence the "making it hard to copy the link for external use"

Another option is to add a link to the repo in the HOMEPAGE variable

---

_Hello @MrRoy,_

> Another option is to add a link to the repo in the HOMEPAGE variable

Thank you for the suggestion!
I've added `https://github.com/errata-ai/vale` to the `HOMEPAGE`.

Still, it's not clear to me why should I focus on the right part of the sentence but ignore the left part that explicitly sets its applicability to 3 options:
  - `HOMEPAGE`
  - `EGIT_REPO_URI`
  - **the domain** of `SRC_URI`

As shown earlier, it is common practice to use `${PN}` exactly like this.

---

> I've added https://github.com/errata-ai/vale to the HOMEPAGE.

Thanks!

> Still, it's not clear to me why should I focus on the right part of the sentence but ignore the left part that explicitly sets its applicability to 3 options:

I have found myself sometimes reviewing a commit and wanting to visit the GitHub repo behind the package, and having no direct link to the repo, having to copy the `https://github.com/errata-ai/${PN}` URI and having to manually replace the `${PN}` part with the package name. It's a slight annoyance, it makes life easier when there's a direct link you can just copy/paste. But having it in `$HOMEPAGE` is a good compromise.

> As shown earlier, it is common practice to use ${PN} exactly like this.

Yes, it's a common practice and I don't think devs will refuse to merge a PR solely due to this.

I merged this PR in cd3e1061a5166265c75fefd81949e42fee2cef3d

---

I find myself fascinated by the suggestion that it's so important to use ${PN} in SRC_URI that you'd be willing to add an additional static copy of the GitHub repository base url *just* to make up for the inability to use the url statically from SRC_URI.

---

I believe that having a repository url in the `HOMEPAGE` is preferable since it's easily accessible with [eix](https://github.com/vaeth/eix).
```shell
$ eix vale$
[I] app-text/vale
     Available versions:  (~)2.28.3^st[2] (~)3.4.1^st[2] (~)3.4.2^st[2] (~)3.5.0^st[1] (~)3.5.0^st[2]
     Installed versions:  3.5.0^st[1](12:32:18 AM 06/14/2024)
     Homepage:            https://vale.sh/docs/ https://github.com/errata-ai/vale
     Description:         A syntax-aware linter for prose built with speed and extensibility in mind

[1] "mim" /usr/overlay
[2] "guru" /var/db/repos/guru
```

---

I'm glad this new version got merged! For those who are still interested, I want get back to determining the percentage of ebuilds that do or do not use `${PN}` in the part of the URL that links to the project. 
> A bit more stats for [Gentoo ebuild repository](https://github.com/gentoo/gentoo):
> 
> ```shell
> gentoo (master|✔) $ ag 'SRC_URI=.*/\${PN}/' | wc -l  # only lines like https://github.com/errata-ai/${PN}/archive/v${PV}.tar.gz
> 5677
> gentoo (master|✔) $ ag 'SRC_URI=.*\${PN}' | wc -l  # total
> 6340
> ```
I'm not convinced that this is the right comparison to make. What we want to compare is the number of ebuilds, where the leading portion of the URL in `^SRC_URI=.*` that represents the project contains `${PN}`, versus the number of ebuilds, where it does not.

First, I sought to identify the most common domains, so that I can scan those for the interesting patterns later.
```sh
$ git rev-parse HEAD
3e3b6279bd902e90032243148e7c96cc755afc6d
$ git grep -Eo '^SRC_URI=.*://[^/]+/' **/*.ebuild| cut -d : -f 1 --complement | sort | uniq -c | sort -g | tail -n 5
    271 SRC_URI="mirror://gentoo/
    337 SRC_URI="https://dev.gentoo.org/
    419 SRC_URI="https://hackage.haskell.org/
   1398 SRC_URI="https://downloads.sourceforge.net/
   6524 SRC_URI="https://github.com/
```
I chose to focus on the two most popular forges. The common patterns for GitHub and SourceForge projects seem to be:
```
https://github.com/[entity]/[project]
https://downlods.sourceforge.net/[project]
https://downlods.sourceforge.net/project/[project]
```
Here's the of ebuilds that contain any variation of `${PN}` within these patterns versus the total number that matches these patterns:
```sh
$ git grep -Eo '^SRC_URI="https://(github.com|downloads.sourceforge.net)/([^/]+/){1,2}' **/*.ebuild | grep -E '\${?PN}?' | wc -l
3030
$ git grep -Eo '^SRC_URI="https://(github.com|downloads.sourceforge.net)/([^/]+/){1,2}' **/*.ebuild | wc -l
7915
```
In conclusion, I find that about about 40% of ebuilds that are hosted on either GitHub or SourceForge use `${PN}` in the relevant part of the URL in `SRC_URI`. That's still a lot but nowhere near your 89.54%.
Keep in mind that ebuilds that match the above mentioned patterns only make up about 55% of all ebuilds that match `^SRC_URI=.*://[^/]+/`, so these numbers may not be totally accurate either.